src/compute-client/src/as_of_selection.rs

```diff
@@ -0,0 +1,521 @@
+        let mut collections = BTreeMap::new();
+        for dataflow in dataflows {
+            let storage_inputs: Vec<_> = dataflow.source_imports.keys().copied().collect();
+            let compute_inputs: Vec<_> = dataflow.index_imports.keys().copied().collect();
+
+            let bounds = match dataflow.as_of.clone() {
+                Some(frontier) => AsOfBounds::single(frontier),
+                None => AsOfBounds::default(),
+            };
+            let bounds = Rc::new(RefCell::new(bounds));
+
+            for id in dataflow.export_ids() {
+                let collection = Collection {
+                    storage_inputs: storage_inputs.clone(),
+                    compute_inputs: compute_inputs.clone(),
+                    bounds: Rc::clone(&bounds),
+                };
+                collections.insert(id, collection);
+            }
+        }
+
+        Self {
+            collections,
+            storage_collections,
+        }
+    }
+
+    /// Returns the state of the identified collection.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the identified collection doesn't exist.
+    fn expect_collection(&self, id: GlobalId) -> &Collection<T> {
+        self.collections
+            .get(&id)
+            .unwrap_or_else(|| panic!("collection missing: {id}"))
+    }
+
+    /// Applies the given as-of constraint to the identified collection.
+    ///
+    /// Returns whether the collection's as-of bounds where changed as a result.
+    fn apply_constraint(&self, id: GlobalId, constraint: Constraint<T>) -> bool {
+        let collection = self.expect_collection(id);
+        let mut bounds = collection.bounds.borrow_mut();
+        match constraint.apply(&mut bounds) {
+            Ok(changed) => {
+                if changed {
+                    info!(%id, %bounds, reason = %constraint.reason, "applied as-of constraint");
+                }
+                changed
+            }
+            Err(changed) => {
+                match constraint.type_ {
+                    ConstraintType::Hard => {
+                        soft_panic_or_log!(
+                            "failed to apply hard as-of constraint \
+                             (id={id}, bounds={bounds}, constraint={constraint:?})"
+                        );
+                    }
+                    ConstraintType::Soft => {
+                        warn!(%id, %bounds, ?constraint, "failed to apply soft as-of constraint");
+                    }
+                }
+                changed
+            }
+        }
+    }
+
+    /// Apply as-of constraints imposed by the frontiers of upstream storage collections.
+    ///
+    /// A collection's as-of _must_ be >= the read frontier of each of its (transitive) storage
+    /// inputs.
+    ///
+    /// Failing to apply this constraint to a collection is an error. The affected dataflow will
+    /// not be able to hydrate successfully.
```

Comment:
In the scope of compute we just return an error. Adapter then just unwraps it, so yes we will panic. Didn't want to spell it out that explicitly because we might decide to handle this differently in the future. For example, you could imaging marking dataflows as "failed" instead.